Wrappers/Python/cil/optimisation/functions/MixedL21Norm.py

```diff
@@ -72,15 +72,24 @@ def _proximal_step_numpy(tmp, tau):
     '''
     # Note: we divide x by tau so the cases of tau both scalar and 
     # DataContainers run
-    tmp /= np.abs(tau, dtype=np.float32)
-    res = tmp - 1
+    try:
+        tmp = np.abs(tau, dtype=np.float32)
+    except np.core._exceptions._UFuncInputCastingError:
+        tmp = tau.abs()
+    
+    
+    arr /= tmp
+    res = arr - 1
     res.maximum(0.0, out=res)
-    res /= tmp
+    res /= arr
+
+    arr *= tmp
```

Comment:
What's the point of this line? `arr` doesn't seem to be used again in this function

Reply:
`arr` is a reference to a `DataContainer` or `ndarray `. We are changing its contents dividing it by `tau` earlier in the function.

This function is meant to return a new `DataContainer`, `res` but it shouldn't change the input parameters. Because `arr` is passed as a reference, after this function has run, it's contents will have changed, which we don't want as that could be used by the caller at a later stage

Reply:
why not create another `tmp_arr`? Are these arrays so large we need to in-place temporarily modify them so we don't run out of memory?

Reply:
Exactly @casperdcl . While running these algorithms we allocate multiple copies of the data and of the target volume. A 2k^3 volume is 32Gb in 32 bits! We don't want that to proliferate.